Poetry install and GitHub actions

tbody-cfs opened the pull request:
Adds a `pyproject.toml` based build using a `pyproject` based build

To install
```
python -m venv .venv
source ./.venv/bin/activate
python -m pip install --upgrade pip
pip install '.[test,build]'
pytest tests
```

This PR also adds Github Actions to automatically build and test the library, and to push tags to pypi.org. For this to work, the repository owner of https://pypi.org/project/fortranformat/ will need to add Github Actions as a trusted publisher. This basically means
1. Go to https://pypi.org/manage/project/fortranformat/settings/publishing/
2. On the Github tab, add owner: `brendanarnold`, repository: `py-fortranformat`, workflow name: `workflow_actions.yml`, environment name: `pypi-publish` and click Add. See [docs](https://docs.github.com/en/actions/security-for-github-actions/security-hardening-your-deployments/configuring-openid-connect-in-pypi)
3. Go to https://github.com/brendanarnold/py-fortranformat/settings/environments. Click `new environment` and add an environment called `pypi-publish` and add maintainers as required reviewers. See [docs](https://docs.github.com/en/actions/managing-workflow-runs-and-deployments/managing-deployments/managing-environments-for-deployment)
4. Make a new tag. This will automatically run the Github Action and then ask if you want to push to pypi

tbody-cfs committed:
Add poetry install

tbody-cfs committed:
Add github actions to automatically test

brendanarnold commented:
Thanks so much @tbody-cfs it’s really appreciated. I’m on leave right now without a computer until next week. When I get back and caught up I’ll take a look

i see [from this](https://packaging.python.org/en/latest/guides/modernize-setup-py-project/#modernize-setup-py-project) that we should definitely have a toml file, although `poetry` is definitely a design choice so would like to investigate a bit first…

thanks again!

tbody-cfs commented:
Thanks for making a very useful project!

That's true. I like `poetry` for neatly tracking the exact versions of the installed dependencies, and for automatically making isolated environments. However, if you'd prefer to use setuptools, I can adjust the `pyproject.toml` for this instead.

Happy to discuss when you're back.

brendanarnold commented:
Hi @tbody-cfs - It looks like `poetry` is a very popular build tool however I'm always a bit wary of introducing dependencies, especially when it isn't immediately clear what should be installed with conda, in a virtualenv, pipx is a sub-dependency and that is installed by `curl`ing a script from the web... (I know there are other ways but still...)

Would a `setuptools` approach be onerous? I presume the big advantage here is that it would work out of the box ...

tbody-cfs committed:
Switch build system to hatchling

tbody-cfs committed:
Merge branch 'brendanarnold:master' into poetry_install_and_github_actions

tbody-cfs committed:
Update README.md for new install and build instructions

tbody-cfs committed:
Fix syntax error

tbody-cfs commented:
Hi @brendanarnold,
I've now updated the PR to use the `hatchling` backend which works out of the box (no extra installs, works out of the box). This PR essentially follow https://packaging.python.org/en/latest/guides/writing-pyproject-toml/ line by line.

To test the GithubActions workflow, I opened a local merge on my fork. You can check this at https://github.com/tbody-cfs/py-fortranformat/pull/1.

brendanarnold commented:
Hey, don't want to mess you around but surely hatchling isn't part of the base Python library and needs to be installed? Hatch needs installing according to https://hatch.pypa.io/latest/install/

According to [this website](https://www.pyopensci.org/python-package-guide/package-structure-code/python-package-build-tools.html#an-ecosystem-of-python-build-tools) `poetry` is by far the most popular build tool after `setuptools` and so if a third party build tool is needed I would prefer poetry. 

Note that the site does recommend not using `setuptools` and so `poetry` looks like the best option - I just wanted to be sure that we weren't installing unnecessary dependencies.

tbody-cfs commented:
Hi @brendanarnold,
No worries — packaging is important to get right!

I've never manually installed `hatchling` myself — it builds out of the box for me, as well as on [mybinder.org](https://mybinder.org/v2/gh/brendanarnold/py-fortranformat/HEAD
). I did need to manually install `poetry`. I'm assuming that hatchling comes bundled with python, but couldn't find a definitive answer online.

Happy to proceed as you prefer. I can use `setuptools`, `poetry` or `hatchling` — whichever you prefer.

brendanarnold commented:
Hi Tom,

So I've downloaded your commits, 'm leaning towards the `poetry` version since it is much more popular and seems to be fairly straightforward. If you OK with that lets push ahead.

Presumably we can remove the `setup.py` in favour of the `pyproject.toml`? We also don't need `twine` since we can use poetry to deploy. These new steps should be documented in `README.md` - happy to help out with these changes.

tbody-cfs committed:
Switch to poetry build

tbody-cfs committed:
Add poetry lock

tbody-cfs committed:
Remove version constraint on poetry and update README

tbody-cfs commented:
Hi Brendan,
I've switched back to using `poetry` and updated the `README`. Let me know what you think :) 
Cheers,
Tom
P.S. I'd recommend squishing the commit history.

tbody-cfs committed:
Py3.8 required by pytest

brendanarnold commented on fortranformat/__init__.py:
```diff
@@ -1,5 +1,3 @@
-__version__ = '2.0.0'
```
Without this, how do I find the version of fortranformat from the repl? This feels like a breaking change ...

brendanarnold replied:
... is it not still common practice to express version of the library in this way?

tbody-cfs replied:
I've added a backwards compatibility solution which will let you
```
import fortranformat
print(fortranformat.__version__)
```
and get the version from pyproject.toml.

From https://stackoverflow.com/questions/67085041/how-to-specify-version-in-only-one-place-when-using-pyproject-toml?rq=3

Poetry doesn't support dynamic versioning (i.e. defining the version in pyproject.toml from something in the library), and this prevents us needing to write the version in two places.

tbody-cfs committed:
Backward compatibility: add __version__ to __init__.py

tbody-cfs committed:
Slightly tidier solution

brendanarnold commented on README.md:
````diff
@@ -81,22 +97,29 @@ make runperformancetests
-
-To upload a version to PyPI run ...
-
 ```
````
```suggestion
```bash
```

brendanarnold commented on README.md:
````diff
@@ -81,22 +97,29 @@ make runperformancetests
 ```
-python setup.py sdist
-twine upload dist/<new version>
+# Build a local distribution
+poetry build
+# Make a new virtual environment
+python -m venv test_env
+# Activate the test_env
+source ./test_env/bin/activate
+# Install the local build into the test_env
+pip install $(find ./dist -name "*.whl")
+# Make and enter a tempdir so import fortranformat doesn't find the fortranformat directory
+mkdir tmp_dir && cd tmp_dir
+# Try run a simple command to make sure that the project installed correctly.
+python -c "from fortranformat import FortranRecordReader as FReader; assert FReader('(2f10.5)').read('1.0000000 2.0000000') == [1.0, 2.0]"
````
```suggestion
# Build a local distribution
poetry build

# Make a new virtual environment
python -m venv test_env

# Activate the test_env
source ./test_env/bin/activate

# Install the local build into the test_env
pip install $(find ./dist -name "*.whl")

# Make and enter a tempdir so import fortranformat doesn't find the fortranformat directory
mkdir tmp_dir && cd tmp_dir

# Try run a simple command to make sure that the project installed correctly.
python -c "from fortranformat import FortranRecordReader as FReader; assert FReader('(2f10.5)').read('1.0000000 2.0000000') == [1.0, 2.0]"
```

brendanarnold commented on README.md:
````diff
@@ -37,10 +37,26 @@ For more detailed usage, see [the guide](https://github.com/brendanarnold/py-for
+```
+Then, start a new terminal window and try run `poetry --version` to make sure that `poetry` has been linked correctly. 
+
+```
+cd path/to/fortranformat
+# Install the project
+poetry install
+# Run pytest and make sure these pass
+poetry run pytest
````
Adding some spacing and syntax highlighting to the shell

```suggestion
```bash
cd path/to/fortranformat

# Install the project
poetry install

# Run pytest and make sure these pass
poetry run pytest
```

brendanarnold reviewed (changes requested):
So I've finally got around to trying it out - happy to merge once suggestions merged in ...

Can you also please update `CHANGELOG.md` to add your contribution? Something like this line in the `[UNRELEASED]` section? ...

```
- Feature: [pull 39](https://github.com/brendanarnold/py-fortranformat/pull/33) Add `poetry` build tool and Github Action for deployment - @tbody-cfs
```

tbody-cfs committed:
Add build tools to CHANGELOG

tbody-cfs committed:
Update README.md

Co-authored-by: Brendan Arnold <brendanarnold@gmail.com>

tbody-cfs committed:
Update README.md

Co-authored-by: Brendan Arnold <brendanarnold@gmail.com>

tbody-cfs committed:
Update README.md

Co-authored-by: Brendan Arnold <brendanarnold@gmail.com>

tbody-cfs commented:
Thanks for the suggestions @brendanarnold. Should be good to go now :)

brendanarnold commented:
Thanks so much @tbody-cfs  for your coperation - its merged
